include/ensmallen_bits/problems/schaffer_function_n1.hpp

```diff
@@ -0,0 +1,68 @@
+/**
+ * @file schaffer_function_n1.hpp
+ * @author Sayan Goswami
+ *
+ * Implementation of Schaffer function N.1.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_PROBLEMS_SCHAFFER_FUNCTION_N1_HPP
+#define ENSMALLEN_PROBLEMS_SCHAFFER_FUNCTION_N1_HPP
+
+namespace ens {
+namespace test {
+
+template<typename MatType = arma::mat>
+class SchafferFunctionN1
+{
+ public:
+  SchafferFunctionN1() : numObjectives(2), numVariables(1)
+  {/* Nothing to do here. */}
+
+  arma::vec Evaluate(const MatType& coords) {
+    arma::vec objectives(numObjectives);
+
+    objectives(0) = std::pow(coords[0], 2);
+    objectives(1) = std::pow(coords[0] - 2, 2);
+
+    return objectives;
+  }
+
+  MatType GetInitialPoint()
+  {
+    return arma::vec(numVariables, 1, arma::fill::randn);
+  }
+
+  size_t NumObjectives() const { return numObjectives; }
+
+  double GetMaximum(size_t objectiveNumber)
```

Comment:
I was thinking of inheriting the Schaffer N.1 function from an abstract base class of multi-objective functions. This ABC can have the `GetMaximum` and `GetMinimum` as methods?

Reply:
Form a users point of view I'd rather do as little as possible to get the optimizer running, which I think would be passing the parameter to the optimizer at construction time. Let me know what you think.

Reply:
Uhm, is the user expected to create a new class (a new multi-objective problem) to use the library?

Reply:
All the problem function we provide are mainly for testing, a user has to implement there own function that implements the necessary methods, as defined in the documentation. An optimizer parameter has the benefit that a user can switch between optimizer without adding new function to an existing problem.

Reply:
So, @zoq we need to add the limits for the individual objectives as well as the objectives to the `Optimize()` call?
@favre49 was suggesting that since we have a new `MultiobjectiveFunctionType`, we can add that to the docs and make sure that the user is aware of the type. What do you say?

Also, apologies I couldn't revert back earlier as I was busy.

Reply:
I'm still not convinced that this should be part of a specific function API (`MultiobjectiveFunctionType`): Bounding constraints aren't multi-objective function specific, more optimizer specific, for example PSO or CMAES use bounding constraints as well:

https://github.com/mlpack/ensmallen/blob/51614c216e43b235b97ff9059f571c8f02e3de37/include/ensmallen_bits/pso/pso.hpp#L91-L92

Are you aware of any library where this is part of the function? I looked into some but couldn't see anything related. But maybe you found something?

Reply:
Oh okay, I get you now! :)
I am making the necessary changes.

Reply:
Okay, I'll check some more libraries as well, maybe there is something we should add, that is currently missing.

Reply:
@zoq I recently came across [this](https://github.com/jMetal/jMetalPy/). 
The files of interest are the [nsga2 optimizer](https://github.com/jMetal/jMetalPy/blob/6fae0a31159873a4dc27f1ea4d86af03b8371733/jmetal/algorithm/multiobjective/nsgaii.py#L37) and the [zdt problem](https://github.com/jMetal/jMetalPy/blob/6fae0a31159873a4dc27f1ea4d86af03b8371733/jmetal/problem/multiobjective/zdt.py#L33) implementation.

Reply:
Marcus, I have added the bounds as optimizer parameters. Please have a look! :)